Flatten DTK by getting rid of subpackages

aprokop opened the pull request.

aprokop commented:
It passes tests except coverage. Any brave soul to review this one?

dalg24 commented on packages/Benchmarks/HybridTransport/src/CMakeLists.txt:
```diff
@@ -29,6 +19,7 @@ TRIBITS_ADD_LIBRARY(
   dtk_hybridtransport
   HEADERS ${HEADERS}
   SOURCES ${SOURCES}
+  DEPLIBS dtk_utils
```
I guess we have to register these dependency by hands now :/

aprokop replied:
Yes, we need to establish the order of compiling, and `DEPLIBS` is the way to go in TriBITS. It is possible that I missed some of them.

aprokop commented on packages/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+ADD_SUBDIRECTORY(Discretization)
+ADD_SUBDIRECTORY(Meshfree)
+ADD_SUBDIRECTORY(MapFactory)
+ADD_SUBDIRECTORY(Benchmarks)
```
Yes, you would do
```
IF(FEATURE)
  ADD_SUBDIRECTORY(Benchmarks)
ENDIF()
```

dalg24 replied:
Sure, but say I disable `Interface`, what happens with `MeshFree` and `Discretization` that currently depend on it?  I don't think we have an equivalent to the enable/disable logic from TriBITS any more.

aprokop commented on packages/Search/examples/bvh_driver/CMakeLists.txt:
```diff
@@ -11,7 +11,7 @@ TRIBITS_ADD_EXECUTABLE(
   ADDED_EXE_TARGET_NAME_OUT bvh_exe_target_name
   )
-TARGET_LINK_LIBRARIES(${bvh_exe_target_name} benchmark::benchmark)
+TARGET_LINK_LIBRARIES(DataTransferKit_bvh benchmark::benchmark ${Trilinos_LIBRARIES})
```
Sorry, I was copying from the larger PR, and did not notice this snafu.

aprokop commented on packages/Search/examples/distributed_tree_driver/CMakeLists.txt:
```diff
@@ -6,6 +6,7 @@ TRIBITS_ADD_EXECUTABLE(
   distributed_tree
   SOURCES distributed_tree_driver.cpp
   )
+TARGET_LINK_LIBRARIES(DataTransferKit_distributed_tree ${Trilinos_LIBRARIES})
```
Because I wasn't careful enough :( will remove.

sslattery commented:
Nice work Andrey. Am I reading this correctly that we still make different libraries for each component but now we have a single package instead of sub packages?

aprokop commented:
@sslattery Correct.

aprokop commented:
I cleaned up the PR.

I also added another patch to it that maybe slightly controversial. It unifies the style of `*_ENABLE_WHATEVER` to produce `HAVE_DATATRANSFERKIT_WHATEVER`. So, instead of `HAVE_DTK_DBC` we'll have `HAVE_DATATRANSFERKIT_DBC` now.

@sslattery @dalg24 Please let me know if you have concerns about the later patch.

dalg24 commented:
I don't like writing `DATATRANSFERKIT` instead of `DTK`.  I can see we are not always consistent but b8f47b4 does not resolve that issue.  I can think of `HAVE_DTK_BOOST` for instance.  I'd say drop that last commit for now.

aprokop commented:
@dalg24 As an alternative, I could just rename to `HAVE_DTK_EXPLICIT_INSTANTIATION`.

dalg24 commented:
> rename to `HAVE_DTK_EXPLICIT_INSTANTIATION`

Fine by me.

aprokop commented:
Actually, nevermind. Users familiar with Trilinos will be looking for `HAVE_DATATRANSFERKIT_EXPLICIT_INSTANTIATION`, this follows the Trilinos style. I just dropped that commit.

aprokop commented:
Pushed one last fix: I accidentally changed the default for DBC from `${PROJECT_NAME}_ENABLE_DEBUG` to `${PACKAGE_NAME}_ENABLE_DEBUG`.
